app/controllers/application_controller.rb

```diff
@@ -3,11 +3,33 @@ class ApplicationController < ActionController::Base
 
   before_action :configure_permitted_parameters, if: :devise_controller?
 
-protected
+  protected
 
-def configure_permitted_parameters
-  devise_parameter_sanitizer.permit(:sign_up, keys: [:firstname, :lastname, :username, :telephone_number, :address, :provider, :uid])
-end
+  def configure_permitted_parameters
+    devise_parameter_sanitizer.permit(:sign_up, keys: [:firstname, :lastname, :username, :telephone_number, :address, :provider, :uid])
+  end
 
+  def current_cart
+    if current_user
+      if session['cart']
+        visitor_cart = ShoppingCart.find(session['cart'])
+        line_items = visitor_cart.line_items
+        line_items.each do |line_item|
+          line_item.update(shopping_cart: current_user.shopping_cart)
+          line_item.save
+        end
+        visitor_cart.destroy
+        session.delete['cart']
+      end
+      current_user.shopping_cart
+    else
+      unless session['cart']
+        cart = ShoppingCart.new
+        cart.save
+        session['cart'] = cart.id
+      end
+      ShoppingCart.find(session['cart'])
+    end
```

Comment:
If the user is not signed in, create a temporary shopping cart for them and store the shopping cart id in the session hash.